### ISSUE TITLE
Type inference for indexing

### ISSUE THREAD

**Issue by Zoxc** [rust-lang/rust#21649]
This example used to work:
```rust
fn main() {
    let array = [true, false];
    
    for i in 0..1 { 
        println!("{}", array[i]);
    }
}
```

Recently you'll get the error: `error: the trait ``core::ops::Index<i32>`` is not implemented for the type ``[bool]`` [E0277]`

**Comment by edwardw** [comment]
The following works:

```rust
fn main() {
    let array = [true, false];

    for i in range(0, 1) { // you mean range(0, 2) right?
        println!("{}", array[i]);
    }
}
```

So this could be a dup of #21672.

**Comment by pornel** [comment]
I think I've ran into this as well:

```rust
pub fn main() {
	let mut x = [[0f64; 3]; 3];

	for i in 0..3 {
		x[i][0] = 1.0; // the type of this value must be known in this context
	}
}
```

**Comment by edwardw** [comment]
Should have been fixed in #21806.
